Implement the main components of the CPU

majorkik opened the pull request:
- impl a set of opcodes and instructions (without unofficial ones);
- impl a basic NES abstraction that CPU and test work and memory management;
- impl tests for memory management, addressing modes, and checking the operation of some instructions;

majorkik committed:
feat: init rust project structure

majorkik committed:
docs: create readme

majorkik committed:
feat: impl status register

- added flag status check
- tests are implemented

majorkik committed:
refactor: fix naming status of flags

- added functions to enable/disable flags

majorkik committed:
refactor: CPU structure

- tests for checking status flags
- move StatusFlag to cpu.rs

majorkik committed:
docs: update README

- write a brief description of the project

majorkik committed:
feat: NES structure and memory management

- add a basic NES structure that includes a CPU and a 64 KB memory area;
- impl methods for reading/writing to memory;
- impl a method to load ROM (game data);
- impl a test of loading data into ROM;

majorkik committed:
test: write tests to read/write to memory

- write tests for methods of reading/writing 8 and 16 bit values to memory;

majorkik committed:
refactor: delete ROM structure

majorkik committed:
fix: change the beginning of the ROM memory area

- the memory for ROM should start from 0x8000 to 0xFFFF
- use the `wrapping_add` method to prevent going out of bounds

majorkik committed:
test: change the addressing to Little endian when reading 16 bit

majorkik committed:
feat: get an operand by addressing mode

- implement the processing of the basic addressing modes that are common to most instructions;
- write tests;

majorkik committed:
docs: add todo-list

majorkik committed:
docs: fix markup

majorkik committed:
docs: update todo-list

majorkik committed:
feature: byte-to-instruction converter (adc)

majorkik committed:
feat: opcode parsing

majorkik committed:
fix: add missing addressing modes

majorkik committed:
feat: switching flag states

majorkik committed:
docs: move todo's to rust readme

majorkik committed:
style: format code

majorkik committed:
style: add commits

majorkik committed:
feat: impl LD (load to x, y, a) instruction

majorkik committed:
feat: impl ST (store to x, y, a) instructions

majorkik committed:
feat: implement logical instructions

majorkik committed:
fix: updating flags

majorkik committed:
feat: implement inc/dec instructions

majorkik committed:
feat: implement operations for inc/dec memory

majorkik committed:
refactor: methods for update zero/negative flags

- fixed updating of values in registers for ldy/sty/tay operations;

majorkik committed:
feat: implement operations for byte comparison

majorkik committed:
feat: implement the bit operation

majorkik committed:
feat: implement JMP operation

majorkik committed:
feat: implement bit shift operations

majorkik committed:
feat: impl default method for CPU

majorkik committed:
feat: implement main loop

- implemented instruction processing inside the loop;

majorkik committed:
feat: implement BRK instruction

majorkik committed:
feat: use move opcodes by refenrece

majorkik committed:
test: LDA instruction

- refactor get_operand_address: use move AddressingMode by reference;

majorkik committed:
test: ST & LD instruction together

majorkik committed:
feat: impl operations for set/clear flags

majorkik committed:
docs: update todo

majorkik committed:
feat: implement branch operations

majorkik committed:
Merge branch 'master' into majorkik/rust-impl

majorkik committed:
feat: add pop/push methods to work with the stack

majorkik committed:
feat: add 16bit pop/push methods to stack

majorkik committed:
feat: implement ADC instruction

majorkik committed:
refactor: add a trait to work with the memory bus

majorkik committed:
feat: implement SBC instruction

majorkik committed:
feat: operations for subroutines and call stack management

bobakxd commented on rust/src/cpu.rs:
```diff
@@ -0,0 +1,1061 @@
+        self.mem_write_8(address, self.cpu.register_y)
+    }
+
+    fn tax(&mut self, opcode: &OpCode) {
```
Is `OpCode` necessary for this instruction?

majorkik replied:
No, I don't need it. I overlooked this because of the use of `#[allow(unused)]`. It was a temporary solution, but now I have removed it.

bobakxd commented on rust/src/cpu.rs:
```diff
@@ -0,0 +1,1061 @@
+    }
+
+    pub fn mem_write_16(&mut self, address: u16, data: u16) {
+        let [high, low] = [(data >> 8) as u8, (data & 0xFF) as u8];
```
Is it possible to just?

```rust
[(data >> 8) as u8, data as u8]
```

majorkik committed:
refactor: unused instructions

- delete `[allow(unused)]`;
- add calls to unused functions;
- add missing instructions to `OpCode::from_bytes`;
